### PR TITLE
Component as Var type

### PR DESCRIPTION
**Pull request by masenf**
@adhami3310 did most of the initial research for this one 👏 

* It depends on importing components from a CDN instead of getting them from the local webpack bundle
* The component var must start with `comp_` to be considered a dynamic component. In the future a special type of computed var will be used as the marker.
* The returned JS code is imported using a dynamic data uri module... but this means the function that evals the module needs to be async so it can `await` the import. Because of this, we cannot have the component function in the initial state (yet).
* Eventually the `make_component` and `component_var` decorator should become part of the framework instead of defined in the sample code.
* Embedded css props are not properly converted because these require the babel emotion plugin... but that plugin does not seem compatible with the standalone babel as it tries to bring in the node.js `fs` module.

## Sample code

```python
import reflex as rx
from reflex.compiler import templates, utils


def make_component(component: rx.Component) -> str:

    rendered_components = {}
    # Include dynamic imports in the shared component.
    if dynamic_imports := component._get_all_dynamic_imports():
        rendered_components.update(
            {dynamic_import: None for dynamic_import in dynamic_imports}
        )

    # Include custom code in the shared component.
    rendered_components.update(
        {code: None for code in component._get_all_custom_code()},
    )

    rendered_components[
        templates.STATEFUL_COMPONENT.render(
            tag_name="MySSRComponent",
            memo_trigger_hooks=[],
            component=component,
        )
    ] = None

    imports = {}
    for lib, names in component._get_all_imports().items():
        if not lib.startswith((".", "/")) and not lib.startswith("http") and not lib == "react":
            imports[f"https://cdn.jsdelivr.net/npm/{lib}" + "/+esm"] = names
        else:
            imports[lib] = names

    module_code_lines = templates.STATEFUL_COMPONENTS.render(
        imports=utils.compile_imports(imports),
        memoized_code="\n".join(rendered_components),
    ).splitlines()[1:]

    # Rewrite imports from `/` to destructure from window
    for ix, line in enumerate(module_code_lines[:]):
        if line.startswith("import "): 
            if 'from "/' in line:
                module_code_lines[ix] = line.replace("import ", "const ", 1).replace(" from ", " = window[", 1)  + "]"
            elif 'from "react"' in line:
                module_code_lines[ix] = line.replace("import ", "const ", 1).replace(' from "react"', " = window.React", 1)
        if line.startswith("export function"):
            module_code_lines[ix] = line.replace("export function", "export default function", 1)
    return "\n".join(module_code_lines)


def component_var():
    def outer(fn):
        def inner(*args, **kwargs) -> str:
            return make_component(fn(*args, **kwargs))
        inner.__name__ = fn.__name__
        return rx.var(inner)
    return outer



class State(rx.State):
    count: int = 0

    def increment(self):
        self.count += 1

    @component_var()
    def comp_buttoned_count(self) -> rx.Component:
        return rx.button(
            rx.text("Hi Mom"),
            f"Click me {self.count}",
            on_click=type(self).increment,
        )


def index() -> rx.Component:
    return (
        rx.box(
            rx.button(
                "Click me",
                on_click=State.increment,
            ),
            rx.text(State.count),
            State.comp_buttoned_count,
        ),
    )


app = rx.App()
app.add_page(index)
```

### CODE REVIEW

**Comment by adhami3310** [comment]
```python
import reflex as rx


class State(rx.State):
    component_string: str = "rx.button('hi')"

    message: rx.Component = rx.el.div("Hello")

    @rx.var
    def component(self) -> rx.Component:
        try:
            comp = eval(
                self.component_string,
                {
                    "rx": rx,
                    "State": State,
                },
            )
            if not isinstance(comp, rx.Component):
                return rx.el.div("Invalid component")
            return comp
        except Exception as e:
            return rx.el.div(str(e))

    def im_feeling_lucky(self):
        import random

        components = [
            rx.el.div("Hello"),
            rx.button("Click me", on_click=State.im_feeling_lucky),
            rx.text_area("Type here"),
            rx.badge("New"),
        ]

        self.message = random.choice(components)


def index():
    return rx.hstack(
        rx.vstack(
            rx.heading(
                "Input",
                size="3",
            ),
            rx.text_area(
                value=State.component_string,
                on_change=State.set_component_string,
                width="100%",
                flex="1",
                font_family="monospace",
            ),
            height="100%",
            flex="1",
        ),
        rx.vstack(
            rx.heading(
                "Output",
                size="3",
            ),
            rx.card(State.component, width="100%", flex="1"),
            rx.hstack(
                rx.button("I'm feeling lucky", on_click=State.im_feeling_lucky),
                State.message,
            ),
            height="100%",
            flex="1",
        ),
        height="100vh",
        width="100vw",
        box_sizing="border-box",
        padding="1rem",
        gap="1rem",
    )


app = rx.App()
app.add_page(index, "/")
```

File: reflex/state.py [hunk 1]
```diff
@@ -1851,8 +1874,76 @@ def get_value(self, key: str) -> Any:
         Returns:
             The value of the field.
         """
+        from reflex.components.component import Component
+
+        def make_component(component: Component) -> str:
```
Comment (masenf) [inline reflex/state.py:1879]: this needs to move out to its own module i think

File: reflex/vars/base.py [hunk 2]
```diff
@@ -2537,3 +2556,38 @@ def _extract_var_data(value: Iterable) -> list[VarData | None]:
     "set_state": "_var_set_state",
     "deps": "_deps",
 }
+
+
+def eval_component(var_name: str) -> Tuple[str, VarData]:
```
Comment (masenf) [inline reflex/vars/base.py:2561]: this should probably end up in the same module as `make_component`

maybe in `reflex/components/dynamic.py`?

**Review by Lendemor — COMMENTED** [review]
Got a small question: 
How does this impact the initial size for the frontend ?

File: reflex/.templates/jinja/web/pages/_app.js.jinja2 [hunk 3]
```diff
@@ -7,6 +7,10 @@ import '/styles/styles.css'
 {% block declaration %}
 import { EventLoopProvider, StateProvider, defaultColorMode } from "/utils/context.js";
 import { ThemeProvider } from 'next-themes'
+import * as React from "react";
+import * as utils_context from "/utils/context.js";
+import * as utils_state from "/utils/state.js";
+import * as radix from "@radix-ui/themes";
```
[the remark this answers is not in the document]
Comment (adhami3310) [reply reflex/.templates/jinja/web/pages/_app.js.jinja2:13]: Yes. In this PR we are not using radix that is stored in the `window`. I assume loading it there takes some additional time (sadly, not sure if there's any way of doing this dynamically).

**Review by picklelo — COMMENTED** [review]
Functionality is working well for me, this is going to be huge.

* In a follow up we need to add more unit tests for all the new functions introduced here
* [discussed offline] There's a limitation on returning vars from other vars (leading to a "bug in React" error) - we should try to catch that happening and prevent it

Once this is rebased it should be good to merge!

File: reflex/components/component.py [hunk 4]
```diff
@@ -2310,3 +2298,6 @@ def create(cls, *children, **props) -> Component:
                 update={"disposition": MemoizationDisposition.ALWAYS}
             )
         return comp
+
+
+load_dynamic_serializer()
```
[the remark this answers is not in the document]
Comment (adhami3310) [reply reflex/components/component.py:2299]: This has to happen before the state is initialized and after the components has initialized, so there's not much time in between.

File: reflex/components/dynamic.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        ] = None
+
+        imports = {}
+        for lib, names in component._get_all_imports().items():
+            if (
+                not lib.startswith((".", "/"))
+                and not lib.startswith("http")
+                and lib != "react"
+            ):
+                imports[f"https://cdn.jsdelivr.net/npm/{lib}" + "/+esm"] = names
```
[the remark this answers is not in the document]
Comment (adhami3310) [reply reflex/components/dynamic.py:55]: I moved it into a function instead, constants doesn't seem appropriate given the thing with f string

File: reflex/components/dynamic.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+            else:
+                imports[lib] = names
+
+        module_code_lines = templates.STATEFUL_COMPONENTS.render(
+            imports=utils.compile_imports(imports),
+            memoized_code="\n".join(rendered_components),
+        ).splitlines()[1:]
+
+        # Rewrite imports from `/` to destructure from window
+        for ix, line in enumerate(module_code_lines[:]):
```
[the remark this answers is not in the document]
Comment (adhami3310) [reply reflex/components/dynamic.py:65]: This code is tied heavily to the code that it precedes, as such the two cannot be separated meaningfully. One thing we can do is make this logic handle general cases of what we import and what we make available in window to make this code less ugly.

**Review by picklelo — APPROVED** [review]
The first step of the new reflex!

**Pull request merged**